Test load and save Feature

georgreen opened the pull request.

Commit:
Ft: Add Function spection
- Added function contract to load_state and save_state
- Described it's expected input and out put

Commit:
Ft: Implemeted unit tests for save_state

Commit:
Chore: Refactored test to improve coverage

Commit:
Chore: Implemented Specification for load_state
- Implemented Contract for specification

Commit:
Ft: Implemented save_state tests
- Implemented Basic tests for save state functionality

Commit:
Ft: Implemented load state tests
- Added basic tests to test load state functionality

Commit:
Chore: Refactored load_state setup method
- Fixed setup dependancy which caused duplicate error when creating new dojo facillity.

Commit:
Chore : Added Mock data in file.txt

Commit:
Ft: Refactored test and fixed setup

Commit:
Ft: Refactored test and fixed setup

ianoti commented on tests/test_load_state.py:
```diff
@@ -1,2 +1,67 @@
+    dojo = model.Dojo("Andela-Kenya1")
+    dojo1 = model.Dojo("Andela-Kenya2")
+    def setUp(self):
+        self.dojo = LoadStateTestCase.dojo
```
this import statement should be `dojo.LoadTestCase` also there is no `LoadTestCase`method in your model file.

ianoti commented on tests/test_load_state.py:
```diff
@@ -1,2 +1,67 @@
+        new_office = model.Office("new_office")
+        new_livingspace = model.LivingSpace("new_livingspace")
+
+        self.dojo1 = LoadStateTestCase.dojo1
```
same comment as above, this import statement is wrong.

ianoti commented on tests/test_load_state.py:
```diff
@@ -1,2 +1,67 @@
+
+    def test_loads_give_path_existing(self):
+        self.dojo.load_state("./db/sqlite.db")
+        self.assertTrue(self.dojo.loaded)
```
you haven't defined `self.dojo.loaded`

georgreen replied:
Yes, I have to implement that after writing the failing tests.

ianoti commented on tests/test_load_state.py:
```diff
@@ -1,2 +1,67 @@
+        with self.assertRaises(DBAlreadyLoadedException):
+            self.dojo.load_state("./db/already_loaded.db")
+
+    def test_handles_large_file(self):
```
the test in line 63 seems unnecessary.

ianoti commented on tests/test_save_state.py:
```diff
@@ -1,2 +1,51 @@
+        self.dojo.save_state()
+        self.assertTrue(os.path.exists("./db/default.db"))
+
+    def test_throw_update_error_if_exists(self):
```
the purpose of this test on line 38 is not too clear.

georgreen replied:
I did this to give the user an option to update data in a DB or overwrite the DB, and the way I intend to implement this is by throwing an update exception if DB already exists, then I handle the exception by prompting the user if they want to overwrite or simply update.

ianoti commented on tests/test_save_state.py:
```diff
@@ -1,2 +1,51 @@
+        with self.assertRaises(TypeError):
+            self.dojo.save(1234567)
+
+    def test_overwrite_existing_db(self):
```
the overwrite parameter seems unnecessary, the library you use to make the db can create and overwrite the db, by creating the table if it doesn't exist and overwriting it if it does.

ianoti reviewed (changes requested):
A better idea to test the save and load state would be to perform the test in one file for both loading the state and saving the state. You can test the load state by loading the application state state stored using the save state and checking that the stored items are the same as the ones earlier saved.

jkanyua commented on .coverage:
```diff
@@ -0,0 +1 @@
+!coverage.py: This is a private format, don't read it directly!{"lines":{"/Users/georgreen/.virtualenvs/amityenv/lib/python3.6/site-packages/nose/plugins/debug.py":[40,41,42,43],"/Users/georgreen/.virtualenvs/amityenv/lib/python3.6/site-packages/nose/plugins/deprecated.py":[40,42,43,44],"/Users/georgreen/.virtualenvs/amityenv/lib/python3.6/site-packages/nose/plugins/doctests.py":[188,189,190,191, ...
```
Please include coverage in gitignore

georgreen replied:
done

georgreen replied:
@andela-jkanyua The build is failing because we have to implement failing tests first then afterward implement the functionality which will make them pass, we have been instructed to follow the TDD development process.

jkanyua reviewed (commented):
@georgreen Travis build is failing.. The log will give you clues on what you need to change for build to pass. Can see a few..

Commit:
Chore: updated gitingore file

Commit:
Chore: updated travis.yml file

Commit:
Ft: Refactored logic to improve readability

Commit:
Feat: Update models, Add support for sqlachemy
- Add save_state functionality on dojo
- Add load_state functionality on gojo
- Refactored code in models to enable database support

Commit:
Ft: Implemented and Updated tests for functionality
- Refactored test for project
- Implemeted test for save and load state
- Refactored tests for reallocation functionality

Commit:
Chore: deleted unnecessary files

Commit:
Chore: updated gitignore

Commit:
Chore: updated coverage configuration file

Commit:
Feat: Implemented ui for save_state and load_state
- Implemeted ui for the db faetures

Commit:
Chore: Fixed issues in Ui

Commit:
Chore: Update readme

Commit:
Chore: Update

Commit:
Chore: remove unnecessary file

Commit:
Feat: Hook Database Feature to UI
- Linked Load_state and save_state to UI

Commit:
Chore: Added assets (screen shoots demos)

bmulobi commented on models/facility.py:
```diff
@@ -130,11 +140,105 @@ def remove_staff(self, old_staff):
 
+    def save_state(self, database_name="default.db", over_write=False, up=" "):
+        if type(database_name) != str:
+            raise TypeError
```
use a more elegant way to handle errors so that the program does not terminate but gives the user a relevant message

bmulobi commented on models/facility.py:
```diff
@@ -130,11 +140,105 @@ def remove_staff(self, old_staff):
+                os.remove("models/database/" + database_name)
+            raise OverWriteException
+        if os.path.exists("models/database/" + database_name) and up:
+            raise UpdateException
```
I can see in this script you raise various exceptions which is ok but they should not terminate the program they should instead give a relevant message to the user so that they can perform any remedial actions that you could explicitly prescribe

bmulobi commented on views/template.py:
```diff
@@ -1,11 +1,17 @@
 # -*- coding: utf-8 -*-
 room_created_message = "A %s called %s has been successfully created!"
```
the C style format specifiers (e.g %s) are deprecated in python 3, try to use the format() method instead
